### PR TITLE
Cacheable broker epoch

### PR DESCRIPTION
**Pull request by gitlw**
Adding the field maxBrokerEpoch to newer versions of the 3 controller requests:
LeaderAndIsrRequest, UpdateMetadataRequest, and StopReplicaRequest.
The purpose is to make the UpdateMetadataRequest cacheable on the controller.

In the current code, when the controller sends out a UpdateMetadataRequest, it needs to instantiate
many copies of the request, with each one targeting a different broker, and thus having a different brokerEpoch value.

This PR will change the 3 controller requests, including the UpdateMedataRequest,
to be using the maximum broker epoch across the cluster,
such that all brokers can share the same object. As a result, the memory footprint
on the controller should be reduced by a factor of O(number of brokers in the cluster).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by radai-rosenblatt** [comment]
is this making it upstream, or will this be a linkedin-specific optimization ?

File: build.gradle [hunk 1]
```diff
@@ -287,6 +287,11 @@ subprojects {
       exceptionFormat = testExceptionFormat
     }
     logTestStdout.rehydrate(delegate, owner, this)()
+    failFast = true
```
Comment (radai-rosenblatt) [inline build.gradle:290]: is there an upstream bug for this?

**Comment by radai-rosenblatt** [comment]
assuming maxBrokerEpoch = ZK epoch when newest broker was created, why not use highest ZK epoch controller has "consumed"? (would be > maxBrokerEpoch and would give a more accurate estimate of the data upon which any controller commands are based?)

File: build.gradle (hunk 1 above)
Comment (gitlw) [reply build.gradle:290]: 1. There are a bunch failures in the security related tests, and I'm excluding them below.
It seems they only fail on our dev machines, and not on the github CI machines. And for now, no one wants to invest enough time to fix them.
2. the failFast setting means stop the testing when the first failure happens, which has always worked more nicely for me.

**Comment by gitlw** [comment]
> is this making it upstream, or will this be a linkedin-specific optimization ?

I'll push for this change in upstream. But that may take a long time, and I don't want that to delay our optimization.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 2]
```diff
@@ -240,11 +240,11 @@ class KafkaApis(val requestChannel: RequestChannel,
     val updateMetadataRequest = request.body[UpdateMetadataRequest]
 
     authorizeClusterAction(request)
-    if (isBrokerEpochStale(updateMetadataRequest.brokerEpoch())) {
+    if (isBrokerEpochStale(updateMetadataRequest.brokerEpoch(), updateMetadataRequest.maxBrokerEpoch())) {
       // When the broker restarts very quickly, it is possible for this broker to receive request intended
       // for its previous generation so the broker should skip the stale request.
       info("Received update metadata request with broker epoch " +
```
Comment (radai-rosenblatt) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: nit - cant always call it "boker epoch" in logs anymore ?

**Review by radai-rosenblatt — APPROVED** [review]
LGTM

**Pull request merged**